File: frontend/src/engine/normalize.ts
// ============================================================================
// engine/normalize.ts — Chart data adapter: raw source rows → Normalized*.
//
// Covers the terminal's real sources:
//   • GET /api/candles rows: [ts, o, h, l, c, v] with ts in epoch seconds
//   • /api/ws tick JSON
//   • shell candle objects that may already be in seconds or ms
// ============================================================================

import type { NormalizedCandle, NormalizedQuote, NormalizedTick } from './types';

/** Below this, a timestamp cannot be ms (would be 1970); treat as seconds. */
export const MS_THRESHOLD = 1e12;

export function toMs(ts: number): number {
  if (!Number.isFinite(ts)) return ts;
  return ts < MS_THRESHOLD ? ts * 1000 : ts;
}

export function toSec(ts: number): number {
  if (!Number.isFinite(ts)) return ts;
  return ts >= MS_THRESHOLD ? Math.floor(ts / 1000) : Math.floor(ts);
}

/** Normalize one API candle row or object. Returns null when unusable. */
export function normalizeCandle(raw: any): NormalizedCandle | null {
  if (!raw) return null;
  let time: number;
  let open: number, high: number, low: number, close: number;
  let volume: number | undefined;
  if (Array.isArray(raw)) {
    if (raw.length < 5) return null;
    [time, open, high, low, close] = raw;
    volume = raw[5] != null ? Number(raw[5]) : undefined;
  } else {
    time = raw.time ?? raw.ts ?? raw.timestamp;
    open = raw.open;
    high = raw.high;
    low = raw.low;
    close = raw.close;
    volume = raw.volume != null ? Number(raw.volume) : undefined;
  }
  if (time == null || ![open, high, low, close].every((n) => Number.isFinite(Number(n)))) {
    return null;
  }
  const t = toMs(Number(time));
  const o = Number(open), h = Number(high), l = Number(low), c = Number(close);
  // Repair inverted high/low only when both sides are finite (corrupt feeds).
  const hi = Math.max(h, o, c, l);
  const lo = Math.min(l, o, c, h);
  return {
    time: t,
    open: o,
    high: hi,
    low: lo,
    close: c,
    volume: volume != null && Number.isFinite(volume) ? volume : undefined,
  };
}

export function normalizeCandles(rows: any[]): NormalizedCandle[] {
  if (!Array.isArray(rows)) return [];
  const out: NormalizedCandle[] = [];
  for (const r of rows) {
    const c = normalizeCandle(r);
    if (c) out.push(c);
  }
  out.sort((a, b) => a.time - b.time);
  return out;
}

export function normalizeQuote(raw: any): NormalizedQuote | null {
  if (!raw) return null;
  const bid = raw.bid != null && Number.isFinite(Number(raw.bid)) ? Number(raw.bid) : null;
  const ask = raw.ask != null && Number.isFinite(Number(raw.ask)) ? Number(raw.ask) : null;
  if (bid == null && ask == null) return null;
  return {
    bid,
    ask,
    synthetic: !!raw.synthetic,
    ts: raw.ts != null ? toMs(Number(raw.ts)) : undefined,
  };
}

export function normalizeTick(raw: any): NormalizedTick | null {
  if (!raw || raw.price == null || !Number.isFinite(Number(raw.price))) return null;
  return {
    symbol: String(raw.symbol || ''),
    price: Number(raw.price),
    ts: raw.ts != null ? toMs(Number(raw.ts)) : Date.now(),
    bid: raw.bid != null ? Number(raw.bid) : null,
    ask: raw.ask != null ? Number(raw.ask) : null,
    volume: raw.volume != null ? Number(raw.volume) : null,
  };
}

/**
 * Merge a live tick into the candle series (mutates a copy of the last bar
 * or appends a new bucket). Volume accumulates when the tick carries size.
 * Returns a NEW array identity when anything changed (React props compare).
 */
export function mergeTickIntoCandles(
  candles: NormalizedCandle[],
  tick: NormalizedTick,
  timeframeId: string,
  opts: { maxBars?: number; stepSeconds: number }
): NormalizedCandle[] {
  if (!opts.stepSeconds) {
    // tick chart: one bar per print
    const bar: NormalizedCandle = {
      time: toMs(toSec(tick.ts)),
      open: tick.price, high: tick.price, low: tick.price, close: tick.price,
      volume: tick.volume ?? 0,
    };
    const next = candles.concat(bar);
    const max = opts.maxBars ?? 20000;
    return next.length > max ? next.slice(-max) : next;
  }
  if (!candles.length) {
    const b = Math.floor(toSec(tick.ts) / opts.stepSeconds) * opts.stepSeconds;
    return [{
      time: toMs(b),
      open: tick.price, high: tick.price, low: tick.price, close: tick.price,
      volume: tick.volume ?? 0,
    }];
  }
  const out = candles.slice();
  const last = { ...out[out.length - 1] };
  const bucket = Math.floor(toSec(tick.ts) / opts.stepSeconds) * opts.stepSeconds;
  const lastSec = toSec(last.time);
  if (bucket > lastSec) {
    out.push({
      time: toMs(bucket),
      open: tick.price, high: tick.price, low: tick.price, close: tick.price,
      volume: tick.volume ?? 0,
    });
  } else {
    last.close = tick.price;
    last.high = Math.max(last.high, tick.price);
    last.low = Math.min(last.low, tick.price);
    if (tick.volume != null && Number.isFinite(tick.volume)) {
      last.volume = (last.volume ?? 0) + tick.volume;
    }
    out[out.length - 1] = last;
  }
  return out;
}

/** Change vs a reference close (session open / previous candle). */
export function priceChange(
  price: number | null | undefined,
  ref: number | null | undefined
): { change: number | null; changePct: number | null } {
  if (price == null || ref == null || !Number.isFinite(price) || !Number.isFinite(ref) || ref === 0) {
    return { change: null, changePct: null };
  }
  const change = price - ref;
  return { change, changePct: (change / ref) * 100 };
}

File: lse_terminal/market_data/normalize.py
"""Timestamp + payload normalization at the adapter edge."""

from __future__ import annotations

import math
import time
from typing import Any, Iterable, Optional

import pandas as pd

from lse_terminal.contracts.types import CANDLE_COLUMNS

from .types import NormalizedCandle, NormalizedQuote, NormalizedTrade

# Heuristic: values below this are not milliseconds (year 2001 in ms ≈ 9.7e11).
_MS_THRESHOLD = 1e12
_US_THRESHOLD = 1e14
_NS_THRESHOLD = 1e17


def now_ms() -> int:
    return int(time.time() * 1000)


def to_ms(ts: float | int | None) -> Optional[int]:
    """Normalize epoch seconds / ms / µs / ns to milliseconds.

    Returns None for non-finite input. Does not silently coerce ISO strings
    (callers pass numeric provider fields).
    """
    if ts is None:
        return None
    try:
        v = float(ts)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(v):
        return None
    av = abs(v)
    if av >= _NS_THRESHOLD:
        return int(v / 1_000_000)
    if av >= _US_THRESHOLD:
        return int(v / 1_000)
    if av >= _MS_THRESHOLD:
        return int(v)
    # seconds (including sub-second fractions)
    return int(v * 1000)


def to_sec(ts: float | int | None) -> Optional[int]:
    ms = to_ms(ts)
    return None if ms is None else ms // 1000


def _finite(x: Any) -> Optional[float]:
    try:
        v = float(x)
    except (TypeError, ValueError):
        return None
    return v if math.isfinite(v) else None


def normalize_tick(raw: dict, *, source: str = "", receive_ms: int | None = None) -> Optional[NormalizedTrade]:
    """Provider tick dict → NormalizedTrade (price path also carries quote fields).

    The legacy stream shape is {symbol, price, ts, bid?, ask?, volume?}.
    When bid/ask are present a paired quote is returned via normalize_quote.
    """
    if not isinstance(raw, dict):
        return None
    price = _finite(raw.get("price"))
    if price is None or price < 0:
        return None
    ts = to_ms(raw.get("ts"))
    if ts is None:
        ts = receive_ms if receive_ms is not None else now_ms()
    side = raw.get("side") or raw.get("taker_side") or raw.get("aggressor")
    if side is not None:
        side = str(side).lower()
        if side not in ("buy", "sell", "b", "s"):
            side = None
        else:
            side = "buy" if side in ("buy", "b") else "sell"
    tid = raw.get("trade_id") or raw.get("id") or raw.get("tradeId")
    size = _finite(raw.get("volume") if raw.get("volume") is not None else raw.get("size"))
    return NormalizedTrade(
        symbol=str(raw.get("symbol") or ""),
        timestamp_ms=ts,
        price=price,
        size=size or 0.0,
        side=side,
        trade_id=None if tid is None else str(tid),
        source=source or str(raw.get("source") or ""),
    )


def normalize_quote(
    raw: dict, *, source: str = "", receive_ms: int | None = None
) -> Optional[NormalizedQuote]:
    if not isinstance(raw, dict):
        return None
    bid = _finite(raw.get("bid"))
    ask = _finite(raw.get("ask"))
    last = _finite(raw.get("price"))
    if bid is None and ask is None and last is None:
        return None
    ts = to_ms(raw.get("ts"))
    if ts is None:
        ts = receive_ms if receive_ms is not None else now_ms()
    if bid is not None and ask is not None and ask < bid:
        # Invalid book top — drop sides rather than invent a repair.
        bid, ask = None, None
    return NormalizedQuote(
        symbol=str(raw.get("symbol") or ""),
        timestamp_ms=ts,
        bid=bid,
        ask=ask,
        bid_size=_finite(raw.get("bid_size") or raw.get("bidSize")),
        ask_size=_finite(raw.get("ask_size") or raw.get("askSize")),
        last=last,
        source=source or str(raw.get("source") or ""),
        synthetic=bool(raw.get("quote_synthetic") or raw.get("synthetic")),
    )


def normalize_candle_row(
    row: Any, *, symbol: str, timeframe: str, source: str = ""
) -> Optional[NormalizedCandle]:
    """Row as [ts,o,h,l,c,v] or mapping with CANDLE_COLUMNS / time / open…."""
    if row is None:
        return None
    if isinstance(row, (list, tuple)):
        if len(row) < 5:
            return None
        ts, o, h, l, c = row[0], row[1], row[2], row[3], row[4]
        v = row[5] if len(row) > 5 else 0.0
    elif isinstance(row, dict):
        ts = row.get("ts", row.get("time", row.get("timestamp")))
        o, h, l, c = row.get("open"), row.get("high"), row.get("low"), row.get("close")
        v = row.get("volume", 0.0)
    else:
        return None
    tms = to_ms(ts)
    o, h, l, c = _finite(o), _finite(h), _finite(l), _finite(c)
    vol = _finite(v)
    if tms is None or None in (o, h, l, c):
        return None
    if vol is None:
        vol = 0.0
    if vol < 0:
        vol = 0.0  # negative volume is invalid; zero out but quality logs via validate
    # Enforce high >= max(o,c,l) and low <= min(o,c,h) only when inconsistent
    # would be impossible — quality engine flags; here we only fix clear swaps.
    hi, lo = max(h, o, c, l), min(l, o, c, h)
    if hi != h or lo != l:
        # Leave as-is for quality detection; do not silently rewrite OHLC.
        pass
    return NormalizedCandle(
        symbol=symbol,
        timeframe=timeframe,
        timestamp_ms=tms,
        open=o,
        high=h,
        low=l,
        close=c,
        volume=vol,
        source=source,
    )


def normalize_candles_df(
    df: pd.DataFrame, *, symbol: str, timeframe: str, source: str = ""
) -> list[NormalizedCandle]:
    """DataFrame with CANDLE_COLUMNS (ts in seconds) → sorted NormalizedCandle list."""
    if df is None or len(df) == 0:
        return []
    cols = [c for c in CANDLE_COLUMNS if c in df.columns]
    if not {"ts", "open", "high", "low", "close"}.issubset(cols):
        return []
    out: list[NormalizedCandle] = []
    for row in df.itertuples(index=False):
        data = row._asdict() if hasattr(row, "_asdict") else dict(zip(df.columns, row))
        nc = normalize_candle_row(
            [data.get("ts"), data.get("open"), data.get("high"), data.get("low"),
             data.get("close"), data.get("volume")],
            symbol=symbol,
            timeframe=timeframe,
            source=source,
        )
        if nc:
            out.append(nc)
    out.sort(key=lambda c: c.timestamp_ms)
    return out


def normalize_provider_capabilities(provider) -> list[str]:
    from .capabilities import formal_capabilities

    return formal_capabilities(provider)


def feed_kind_from_health(
    health: str, *, has_live: bool, is_replay: bool = False, is_historical: bool = False
) -> str:
    from .types import ConnectionHealth, FeedKind

    if is_replay:
        return FeedKind.REPLAY.value
    if health == ConnectionHealth.CONNECTED.value and has_live:
        return FeedKind.LIVE.value
    if health in (
        ConnectionHealth.RECONNECTING.value,
        ConnectionHealth.CONNECTING.value,
        ConnectionHealth.DEGRADED.value,
    ):
        return FeedKind.DELAYED.value
    if is_historical or not has_live:
        return FeedKind.HISTORICAL.value
    return FeedKind.UNAVAILABLE.value
